QCOW import/export

last-genius opened the pull request:
Depends on https://github.com/xcp-ng-rpms/xs-opam-repo/pull/10

Should also be against some qcow feature branch, not against `v25.6.0-8.3`

gthvn1 committed:
[qcow-tool packaging] import qcow-tool from ocaml-qcow

We just clone the repo without modifications from
https://github.com/mirage/ocaml-qcow.git

This breaks the build...

This is the first step to enable VDI import/export for qcow2 file.
This commit is just the import from Mirage. Next step is to build
the package.

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

gthvn1 committed:
[qcow-tool packaging] build fix: use io-page instead of io-page-unix

Use io-page instead of io-page-unix because io-page is 3.0.0 and
io-page-unix is 2.3.0 only

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

gthvn1 committed:
[qcow-tool packaging] build fix: cstruct

Since 6.0.1 Cstruct.len function was deprecated and we must use
Cstruct.length. This patch fixes this issue

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

gthvn1 committed:
[qcow-tool packaging] build fix: remove Unimplemented variant

In MirageOS, the Unimplemented variant was removed from the
Mirage_block.error type in version 2.0.0. This patch removes
all places where this varient is used. In some places we
replace it with a Failure.

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

gthvn1 committed:
[qcow-tool packaging] build fix: update cmdliner

In recent versions of cmdliner, Term.pure and Term.info have been
replaced with newer, more explicit constructors. This patch uses
Term.const as a replacement.

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

gthvn1 committed:
[qcow-tool packaging] package it in xapi

remove dune-project from `ocaml/qcow-tool` that was here from the first
import. Now we are building qcow-tool with xapi. This patch creates
the package in the toplevel dune-project and removes uneeded build
stuff from the original repository.

To be able to build qcow-tool we added the following packages to xs-opam:
  - asetmap
  - ezjsonm
  - mirage-block-combinators
  - mirage-types (deprecated)
  - mirage-types-lwt (deprecated)
  - prometheus
  - unix-type-representations

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

gthvn1 committed:
[qcow-tool packaging] remove lib-test

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

gthvn1 committed:
[qcow-tool] run make format

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

last-genius committed:
python: Add qcow2-to-stdout.py script

Taken from https://github.com/qemu/qemu/blob/a9cd5bc6399a80fcf233ed0fffe6067b731227d8/scripts/qcow2-to-stdout.py

> This tool converts a disk image to qcow2, writing the result directly
> to stdout. This can be used for example to send the generated file
> over the network.
>
> This is equivalent to using qemu-img to convert a file to qcow2 and
> then writing the result to stdout, with the difference that this tool
> does not need to create this temporary qcow2 file and therefore does
> not need any additional disk space.

Signed-off-by: Andrii Sultanov <andriy.sultanov@vates.tech>

last-genius committed:
python3/qcow2-to-stdout: Update to handle block special files

The original script is designed to only work on regular files. Adapt it
slightly to work on special block devices (since that's what we want to export
from).

Aside from dropping some unnecessary checks and replacing os.path.getsize with
a functional equivalent that works on block special files, these changes remove
detection of "holes" in the raw file, since SEEK_DATA and SEEK_HOLE are
unimplemented in this case.

Signed-off-by: Andrii Sultanov <andriy.sultanov@vates.tech>

last-genius committed:
ocaml-qcow: Update header parsing with the new QCOW fields

Since the library was developed, QCOW added an additional compression_type
field, which can, when present, increase the header size from the expected
minimum 104 to 112 (compression_type only takes up a single byte, with padding
rounding the size up to a multiple of 8).

Qcow_header parsing did not account for additional fields, however, and
expected exactly 104 byte-long headers, raising:

Failure("Read a header_length of 112 but we computed 104")

Since "In general, these fields are optional and may be safely ignored by the
software", do exactly that - expect a header *at least* 104 bytes long with the
size a multiple of 8.

Signed-off-by: Andrii Sultanov <andriy.sultanov@vates.tech>

last-genius committed:
qcow-tool: Fix positional argument parsing in the CLI

Since both "filename" (input file) and "output" arguments were defined by
the positional argument at the 0th position, only the defaults could be used in
the commands that required both, which would not make sense for some commands
(e.g., encode would take test.qcow2 as the input raw file and produce test.raw
as the output qcow2 file).

Whereas if one were to specify `qcow-tool encode test.raw test.qcow2`, both the
input and the output file would be test.raw, and the 1st positional argument
would go unused.

Fix this by making the output argument take the argument at 1st position, and
provide specific per-command defaults.

Signed-off-by: Andrii Sultanov <andriy.sultanov@vates.tech>

last-genius committed:
ocaml-qcow: Fix malloc page allocation calculation

A page is 1<<12 (4096) bytes. A cluster is 1<<cluster_bits bytes.
A single cluster fits 1<<(cluster_bits-12) pages, not
cluster_bits<<(cluster_bits-12) pages.

This error would allocate cluster_bits more pages than needed. For a 16-bit
cluster_bits value (65536 bytes), for example, malloc would allocate 256=16*16
pages instead of the required 16.

Signed-off-by: Andrii Sultanov <andriy.sultanov@vates.tech>

last-genius committed:
qcow-tool: Add qcow_stream module implementation - with debug logging

Separate it into its own library with minimal dependencies.

Signed-off-by: Andrii Sultanov <andriy.sultanov@vates.tech>

last-genius committed:
cli: Add stream_decode command

Signed-off-by: Andrii Sultanov <andriy.sultanov@vates.tech>

last-genius committed:
{export,import}_raw_vdi: add qcow as supported format

This patch allows to pass "qcow2" as a supported format when calling VDI
export and import.

Qcow_tool_wrapper is added as a helper that calls the Python script for export
(conversion from raw to qcow2 stream) and the OCaml Qcow_stream library for
import (conversion from qcow2 stream to raw).

vdi-import with format=qcow2 also supports not specifying the VDI (just like
format=raw). In this case, xapi will read the QCOW2 header first to determine
the virtual size and create the VDI itself before copying over the data.

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>
Signed-off-by: Andrii Sultanov <andriy.sultanov@vates.tech>

last-genius committed:
tests/unix_select: Add Lwt_engine__fun to the list of expected functions calling select

Signed-off-by: Andrii Sultanov <andriy.sultanov@vates.tech>

last-genius committed:
ocaml-qcow: Skip and cache subsequent clusters when reading non-sequentially

Turns out, various tools do tend to produce less nice QCOW files, where the
headers->refcount tables->L1 tables->L2 tables are not all laid out
sequentially cluster-after-cluster, without any gaps in the middle.
Even qcow2-to-stdout and qemu-img can sometimes leave a gap between the
refcount cluster and the first L1 table cluster, for some reason.

So we need to handle non-sequential clusters when reading - modify Qcow_cache
so that it knows how to handle non-seekable input in the case where several
clusters need to be skipped over (and cached) and only the last one needs
returning immediately.

This way we can return the skipped-over clusters from the cache later if they
are needed.

Signed-off-by: Andrii Sultanov <andriy.sultanov@vates.tech>

last-genius commented:
CI will be failing because a custom `xs-opam` is required.

last-genius committed:
python3/qcow2-to-stdout: Add --diff option to export only changed clusters

When specified, only the clusters that have changed between diff_file_name and
input_file will be allocated and exported in a sparse manner.

This is analogous to vhd-tool's --relative-to option.

Example usage:
```
Create an empty raw file:
$ dd if=/dev/zero of=disk.before.img bs=1M count=100

Fill the first three clusters with random data
$ dd if=/dev/random conv=notrunc of=disk.before.img bs=65536 count=3
$ python3/libexec/qcow2-to-stdout.py disk.before.img > disk.before.qcow2

Check that only the first three clusters are allocated (each is 0x10000 in length)
$ qemu-img map disk.before.qcow2
Offset          Length          Mapped to       File
0               0x30000         0x50000         disk.before.qcow2

Overwrite the 2nd and 3rd clusters with new data
$ cp disk.before.img disk.after.img
$ dd if=/dev/random conv=notrunc of=disk.after.img bs=65536 count=2 oseek=1

Export the difference, verifying that only the 2nd and 3rd clusters are allocated
$ python3/libexec/qcow2-to-stdout.py disk.after.img --diff disk.before.img > disk.diff.qcow2
$ qemu-img map disk.diff.qcow2
Offset          Length          Mapped to       File
0x10000         0x20000         0x50000         disk.diff.qcow2

The image can be recreated if the base is imported first, with the difference
overwriting it
$ qemu-img convert -f qcow2 -O raw disk.before.qcow2 disk.test.img
$ qemu-img convert -f qcow2 -O raw disk.diff.qcow2 disk.test.img --target-is-zero -n
$ diff disk.test.img disk.after.img
<the same>

```

Signed-off-by: Andrii Sultanov <andriy.sultanov@vates.tech>

last-genius committed:
export_raw_vdi: Add support for differential QCOW2 export with base

Signed-off-by: Andrii Sultanov <andriy.sultanov@vates.tech>

gthvn1 commented on ocaml/qcow-tool/lib/qcow_header.ml:
```diff
@@ -423,9 +423,12 @@ let read rest =
-    error_msg "Read a header_length of %d but we computed %d" header_length
-      (sizeof t)
+  (* we ignore additional fields like compression_type *)
+  if sizeof {t with extensions= []} > header_length then
```
And also probably check that header _length is at least 104 for v3. But I guess that sizeof t is always 104 and if I'm right, if header length is below 104 we will catch it...

last-genius replied:
Yes, sizeof = 104

gthvn1 reviewed (approved):
LGTM. I only need to review the diff option added to the python script but otherwise it looks good !!!

last-genius commented:
Fixed a bug where a gap between refcount clusters and L1 clusters wasn't counted against `max_clusters`, leading to `Reference_outside_file` errors: https://github.com/xcp-ng/xen-api/compare/d7bab8308107c3f00bd4e61beaa3e6ea6a101230..6f20b3bb339c9fff88ebcf06d75299d0435244a1

gthvn1 commented:
@last-genius , so I think that we can merge it into a new qcow branch to prepare the SM alpha 3 pre release :). What do you think?

gthvn1 commented:
and we need to do it for the opam repo PR as well.

last-genius commented:
I'll refresh this branch today with the changes from upstream review and can merge it afterwards

last-genius commented:
opam repo PR is already merged, build is here: https://koji.xcp-ng.org/buildinfo?buildID=3864, we can just tag it into the qcow repo

gthvn1 commented:
Yes my point was to make opam repo available for qcow release :+1:

last-genius commented:
Changed the base branch to `feature-qcow`, refreshed the PR with some fixes from upstream review. Will do a quick test to see if anything broke accidentally and we can discuss how to tag the builds.

last-genius committed:
Add qcow2 as supported format by xcp-rrdd-iostat

Since we are now supporting qcow file `tap-ctl list` can return strings like:
  - "1564848    0    0      qcow2 /var/run/sr-mount/..."
Without this patch the type "qcow2" is unknown and xcp-rrdd-iostat generates an
error like: returned a line that could not be parsed. Ignoring
This patch fixes the issue.

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

last-genius commented:
Added a `xcp-rrdd-iostat` backport

last-genius commented:
Tests look good, so I'm merging this as "alpha"
